Add `CovidcastRow` testing util and a few other changes

dshemetov opened the pull request:
I made a number of QoL changes to the tests and the server in the process of working on JIT (see also #949). I decided to separate them out, so we can release them separately from JIT and help debug if anything goes wrong.

Built on #1072.

**Prerequisites**:

- [x] Unless it is a documentation hotfix it should be merged against the `dev` branch
- [x] Branch is up-to-date with the branch to be merged with, i.e. `dev`
- [x] Build is successful
- [x] Code is cleaned up and formatted

### Summary

- Adds CovidcastRow testing dataclass, which serves as an abstraction for a row in the database
  - Supports many constructors (from arguments, from dictionaries)
  - Brings many defaults, so tests only need to specify the fields they care about
  - Has a list class companion CovidcastRows that helps in making a bundle of rows
  - Supports three views as a DataFrame, with consistently enforced data types for each column
- Integrates this class into tests across the codebase
- Integrates this class into `database.py` in acquisition
  - [x] Need to update CovidcastRow to remove default values for safety
- Adds `more_itertools` dependency, as it's needed in a few tests and needed later in JIT
- Adds PANDAS_DTYPES to model.py to synchronize the datatypes used by Pandas in JIT and in CovidcastRow (will remove, if JIT does not end up using Pandas)
- Removes some unused imports in a few server files
- Fixes a Pandas warning due to assigning on a view in test_csv_uploading

krivard commented:
Oh no, #1043 was squash-merged, so now git won't be able to automatically identify any commits from this PR that were also in that one. You'll need to rebase -i and drop the following:
* 6d82286d0f311ea722b6b0ac8c91f20b01166bb0
* 33037dc50348fcd54f427b235e8c7ee56d1e4fad
* 9a3c07d5864a3553fbd0fe141a42ec9f8e390cd6
* 96153dbc5b21eb84e1391e0bb27372d45f08d37b
* 8da9b0feb9c793a73b8a1487bba336da4e533565

dshemetov commented:
True, fortunately I'm ok with rebasing and force pushing into my branches, since no one else develops on these.

krivard commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -10,41 +12,52 @@
-        )
+        if is_compatibility:
+            url = BASE_URL_OLD
+            params.setdefault("endpoint", "covidcast")
```
```suggestion
            # only set endpoint if it's not already set
            params.setdefault("endpoint", "covidcast")
```

krivard commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -10,41 +12,52 @@
         return response.json()
 
+    def _diff_rows(self, rows: Sequence[float]):
+        return [float(x - y) if x is not None and y is not None else None for x, y in zip(rows[1:], rows[:-1])]
```
readability nit
```suggestion
        return [
            float(x - y) if x is not None and y is not None else None
            for x, y in zip(rows[1:], rows[:-1])
        ]
```

krivard commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -53,20 +66,52 @@ def test_basic(self):
+        self._insert_rows(rows)
+
+        with self.subTest("simple"):
+            # TODO: These tests aren't actually testing the compatibility endpoint.
```
> `# TODO: These tests aren't actually testing the compatibility endpoint.`

Should this be resolved before merge?

krivard commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -53,20 +66,52 @@ def test_basic(self):
+            self.assertEqual(len(out["epidata"]), len(rows))
+
+        with self.subTest("unknown signal"):
+            rows = [CovidcastRow(source="jhu-csse", signal="confirmed_unknown", time_value=20200401 + i, value=i) for i in range(10)]
+            first = rows[0]
+            self._insert_rows(rows)
```
How is this different from subTest `simple`? In both cases, it looks like we're querying the data we just inserted. Did you mean to query for something that wasn't available in the db?

dshemetov replied:
Hm, good point. It doesn't look any different except it compares the values. I'll merge these two.

krivard commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,270 @@
+from delphi_utils import Nans
+
+from delphi.epidata.acquisition.covidcast.csv_importer import CsvImporter
+from delphi.epidata.server.utils.dates import day_to_time_value, time_value_to_day
+from delphi.epidata.server.endpoints.covidcast_utils.model import PANDAS_DTYPES
```
🔥 
At first I was like "is that allowed???" but then I looked at the acquisition environment on the server and hot damn it is!

dshemetov replied:
Yea, it seems like all the delphi-epidata code can get server imports, but the server can't import non-server code (mostly because of unnecessarily specific Docker build instructions).

krivard commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,270 @@
+    """
+
+    # Arguments.
+    source: str = "src"
+    signal: str = "sig"
+    time_type: str = "day"
+    geo_type: str = "county"
+    time_value: int = 20200202
+    geo_value: str = "01234"
+    value: float = 10.0
+    stderr: float = 10.0
+    sample_size: float = 10.0
+    missing_value: int = Nans.NOT_MISSING.value
+    missing_stderr: int = Nans.NOT_MISSING.value
+    missing_sample_size: int = Nans.NOT_MISSING.value
+    issue: Optional[int] = 20200202
+    lag: Optional[int] = 0
+    id: Optional[int] = None
+    direction: Optional[int] = None
+    direction_updated_timestamp: int = 0
+    value_updated_timestamp: int = 20200202
```
I feel 😬 about including defaults here; I'm worried they will leak into the actual acquisition stream by mistake. The dedicated builder function (`_make_placeholder_row`) may be ugly, but it does enforce a clear line between what's live data and what's test data. With this, all it would take is for some future change to forget to set one of the 17 fields, and the default will go into the database instead.

dshemetov replied:
Decided to go with option 1 after talking to Katie.

krivard commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,270 @@
+        self.value_updated_timestamp = day_to_time_value(self.value_updated_timestamp) if isinstance(self.value_updated_timestamp, date) else self.value_updated_timestamp
+
+    def _sanity_check_fields(self, extra_checks: bool = True):
+        if self.issue and self.issue < self.time_value:
+            self.issue = self.time_value
```
Do we really want it to correct this silently?

dshemetov replied:
So I don't actually do this by default. It is only done in CovidcastRows.from_args by default, but not in from_records. Admittedly, I should probably just make it off by default everywhere.

krivard commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,270 @@
+        else:
+            self.lag = None
+
+        # This sanity checking is already done in CsvImporter, but it's here so the testing class gets it too.
```
is there a matching note in CsvImporter, so we remember to update it here too if we ever need to change it there?

krivard replied:
Alternately, why not have CsvImporter call this code?

dshemetov replied:
Will make a note.

It would be nice to integrate CsvImporter with this code, agreed. Should probably make a separate PR though.

krivard commented on integrations/client/test_delphi_epidata.py:
```diff
@@ -54,12 +53,12 @@ def test_covidcast(self):
     # insert placeholder data: three issues of one signal, one issue of another
     rows = [
-      self._make_placeholder_row(issue=self.DEFAULT_ISSUE + i, value=i, lag=i)[0]
+      CovidcastRow(issue=20200202 + i, value=i, lag=i)
```
What's the benefit of switching to constant 20200202 here?

dshemetov replied:
Faster to read code when the variables stay local and the abstraction seemed to have no other benefit, especially when I deleted _make_placeholder_row in CovidcastBase (since CovidcastRow already has defaults), it meant I needed to look into CovidcastBase just to find an arbitrary value.

krivard commented on integrations/client/test_delphi_epidata.py:
```diff
@@ -275,7 +275,7 @@ def test_covidcast_meta(self):
     # 3rd issue: 2 12 22
     rows = [
-      self._make_placeholder_row(time_value=self.DEFAULT_TIME_VALUE + t, issue=self.DEFAULT_ISSUE + i, value=t*10 + i)[0]
+      CovidcastRow(time_value=2020_02_02 + t, issue=2020_02_02 + i, value=t*10 + i)
```
Not a fan of replacing with constant `2020_02_02` here, since it has to sync up with lines 302, 303, 309 below

dshemetov replied:
My main preference was having the variable easily viewable nearby and not have to go to CovidcastBase for it. How about making a local DEFAULT_TIME_VALUE, DEFAULT_ISSUE variables in this function instead?

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -10,12 +10,13 @@
 from delphi.epidata.acquisition.covidcast.test_utils import CovidcastBase
 
 # use the local instance of the Epidata API
+# TODO: should we still be using this?
```
> `# TODO: should we still be using this?`

Yes -- all Delphi's pre-pandemic users probably still use the old old delphi-epidata client, which means they probably still hit the api.php endpoint.

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -10,12 +10,13 @@
 BASE_URL = 'http://delphi_web_epidata/epidata/api.php'
-
-
+IGNORE_FIELDS = ["id", "direction_updated_timestamp", "value_updated_timestamp", "source", "time_type", "geo_type"]
```
I have beef with stuffing test materials in CovidcastRow (see below) but if we're going to do it, seems like this array belongs there too, and not as a file constant here

krivard commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -10,41 +12,52 @@
+        return [float(x - y) if x is not None and y is not None else None for x, y in zip(rows[1:], rows[:-1])]
+
+    def _smooth_rows(self, rows: Sequence[float]):
+        return [sum(e)/len(e) if None not in e else None for e in windowed(rows, 7)]
```
same nit

krivard reviewed (commented):
Partial review pass up through ~half of covidcast_row.py

We should probably talk in a synchronous format about how to architect CovidcastRow for use in production vs tests; this complete overlap approach makes me hella nervous

melange396 commented:
looking at this now...  do you wanna base it against `dev`?  that way you might be able to keep the diff on #1072 smaller

melange396 commented on integrations/server/test_covidcast.py:
```diff
@@ -188,10 +181,12 @@ def test_raw_json_format(self):
-    response, expected = self.request_based_on_row(row, **{'format':'json'})
+    response = self.request_based_on_row(row, **{'format':'json'})
+
+    expected = [row.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields)]
```
same as in `integrations/client/test_delphi_epidata.py`, this pattern (`.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields)`) is ripe for shortcutting/reuse

krivard replied:
suggestion: another `as_api_compatibility_dict`

melange396 commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -10,41 +12,58 @@
         """Request a signal from the / endpoint."""
-
-        rows = [self._make_placeholder_row(time_value=20200401 + i, value=i)[0] for i in range(10)]
+        rows = [CovidcastRow.make_default_row(time_value=20200401 + i, value=i) for i in range(10)]
```
nit:
```suggestion
        rows = [CovidcastRow.make_default_row(time_value=2020_04_01 + i, value=i) for i in range(10)]
```

melange396 commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -53,20 +72,35 @@ def test_basic(self):
+            out = self._fetch("/", signal=first.signal_pair, geo=first.geo_pair, time="day:*")
+            self.assertEqual(len(out["epidata"]), len(rows))
+
+    def test_compatibility(self):
```
note: expanding test coverage!  youda bessss!!!!!  🎉

melange396 commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -53,20 +72,35 @@ def test_basic(self):
+
+    def test_compatibility(self):
+        """Request at the /api.php endpoint."""
+        rows = [CovidcastRow.make_default_row(source="src", signal="sig", time_value=20200401 + i, value=i) for i in range(10)]
```
nit:
```suggestion
        rows = [CovidcastRow.make_default_row(source="src", signal="sig", time_value=2020_04_01 + i, value=i) for i in range(10)]
```

melange396 commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -53,20 +72,35 @@ def test_basic(self):
 
         num_rows = 30
-        rows = [self._make_placeholder_row(time_value=20200401 + i, value=i)[0] for i in range(num_rows)]
+        rows = [CovidcastRow.make_default_row(time_value=20200401 + i, value=i) for i in range(num_rows)]
```
nit:
```suggestion
        rows = [CovidcastRow.make_default_row(time_value=2020_04_01 + i, value=i) for i in range(num_rows)]
```

melange396 commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -90,16 +124,17 @@ def test_trend(self):
 
         num_rows = 3
-        rows = [self._make_placeholder_row(time_value=20200401 + i, value=num_rows - i)[0] for i in range(num_rows)]
+        rows = [CovidcastRow.make_default_row(time_value=20200401 + i, value=num_rows - i) for i in range(num_rows)]
```
nit:
```suggestion
        rows = [CovidcastRow.make_default_row(time_value=2020_04_01 + i, value=num_rows - i) for i in range(num_rows)]
```

melange396 commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -185,31 +221,26 @@ def test_correlation(self):
         """Request a signal from the /csv endpoint."""
 
-        rows = [self._make_placeholder_row(time_value=20200401 + i, value=i)[0] for i in range(10)]
+        rows = [CovidcastRow.make_default_row(time_value=20200401 + i, value=i) for i in range(10)]
```
nit:
```suggestion
        rows = [CovidcastRow.make_default_row(time_value=2020_04_01 + i, value=i) for i in range(10)]
```

melange396 commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -185,31 +221,26 @@ def test_correlation(self):
-        issue_0 = [self._make_placeholder_row(time_value=20200401 + i, value=i, sample_size=1, lag=0, issue=20200401 + i)[0] for i in range(num_rows)]
-        issue_1 = [self._make_placeholder_row(time_value=20200401 + i, value=i + 1, sample_size=2, lag=1, issue=20200401 + i + 1)[0] for i in range(num_rows)]
-        last_issue = [self._make_placeholder_row(time_value=20200401 + i, value=i + 2, sample_size=3, lag=2, issue=20200401 + i + 2)[0] for i in range(num_rows)] # <-- the latest issues
+        issue_0 = [CovidcastRow.make_default_row(time_value=20200401 + i, value=i, sample_size=1, lag=0, issue=20200401 + i) for i in range(num_rows)]
+        issue_1 = [CovidcastRow.make_default_row(time_value=20200401 + i, value=i + 1, sample_size=2, lag=1, issue=20200401 + i + 1) for i in range(num_rows)]
+        last_issue = [CovidcastRow.make_default_row(time_value=20200401 + i, value=i + 2, sample_size=3, lag=2, issue=20200401 + i + 2) for i in range(num_rows)] # <-- the latest issues
```
nits:
```suggestion
        issue_0 = [CovidcastRow.make_default_row(time_value=2020_04_01 + i, value=i, sample_size=1, lag=0, issue=2020_04_01 + i) for i in range(num_rows)]
        issue_1 = [CovidcastRow.make_default_row(time_value=2020_04_01 + i, value=i + 1, sample_size=2, lag=1, issue=2020_04_01 + i + 1) for i in range(num_rows)]
        last_issue = [CovidcastRow.make_default_row(time_value=2020_04_01 + i, value=i + 2, sample_size=3, lag=2, issue=2020_04_01 + i + 2) for i in range(num_rows)] # <-- the latest issues
```
theres a lot of these that are all the same date......  🥵

dshemetov replied:
Refactor'd!

melange396 commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -231,7 +262,7 @@ def test_meta(self):
 
         num_rows = 10
-        rows = [self._make_placeholder_row(time_value=20200401 + i, value=i, source="fb-survey", signal="smoothed_cli")[0] for i in range(num_rows)]
+        rows = [CovidcastRow.make_default_row(time_value=20200401 + i, value=i, source="fb-survey", signal="smoothed_cli") for i in range(num_rows)]
```
nit:
```suggestion
        rows = [CovidcastRow.make_default_row(time_value=2020_04_01 + i, value=i, source="fb-survey", signal="smoothed_cli") for i in range(num_rows)]
```

melange396 commented on integrations/server/test_covidcast_endpoints.py:
```diff
@@ -272,22 +303,22 @@ def test_coverage(self):
 
         num_geos_per_date = [10, 20, 30, 40, 44]
         dates = [20200401 + i for i in range(len(num_geos_per_date))]
```
nit:
```suggestion
        dates = [2020_04_01 + i for i in range(len(num_geos_per_date))]
```

melange396 commented on tests/acquisition/covidcast/test_covidcast_row.py:
```diff
@@ -0,0 +1,103 @@
+
+class TestCovidcastRows(unittest.TestCase):
+    def test_transpose_dict(self):
+        assert transpose_dict(dict([["a", [2, 4, 6]], ["b", [3, 5, 7]], ["c", [10, 20, 30]]])) == [{"a": 2, "b": 3, "c": 10}, {"a": 4, "b": 5, "c": 20}, {"a": 6, "b": 7, "c": 30}]
```
nit: split the line for readability

melange396 commented on tests/acquisition/covidcast/test_covidcast_row.py:
```diff
@@ -0,0 +1,103 @@
+            "signal": "sig",
+            "time_type": "day",
+            "geo_type": "county",
+            "time_value": 20200202,
```
nit:
```suggestion
            "time_value": 2020_02_02,
```
im not gonna mark the other ones in this file, do em if ya wanna

melange396 commented on tests/acquisition/covidcast/test_covidcast_row.py:
```diff
@@ -0,0 +1,103 @@
+
+
+    def test_covidcast_rows(self):
+        covidcast_rows = covidcast_rows_from_args(signal=["sig_base"] * 5 + ["sig_other"] * 5, time_value=date_range("2021-05-01", "2021-05-05").to_list() * 2, value=list(range(10)))
```
nit: split line for readability
```suggestion
        covidcast_rows = covidcast_rows_from_args(
            signal=["sig_base"] * 5 + ["sig_other"] * 5,
            time_value=date_range("2021-05-01", "2021-05-05").to_list() * 2,
            value=list(range(10)))
```

melange396 commented on tests/acquisition/covidcast/test_covidcast_row.py:
```diff
@@ -0,0 +1,103 @@
+        }), CovidcastRow._pandas_dtypes)
+        assert_frame_equal(df, expected_df)
+
+        covidcast_rows = covidcast_rows_from_args(signal=["sig_base"] * 5 + ["sig_other"] * 5, time_value=date_range("2021-05-01", "2021-05-05").to_list() * 2, value=list(range(10)))
```
nit: split line for readability

melange396 reviewed (commented):
this is fantastic!  it needed to happen and thank you for taking care of it.  `CovidcastRow` came out nice and OO, i really like it.

but im sorry :( there are a whopping 35 comments on this....  a lot are real easy: 3 are just notes you can dismiss, and 17 are nits (almost all w/ pre-populated suggestions).  i jumped around a lot while reviewing (not in file-order), but hopefully it should all make sense.  the meatier things are: the schema of columns in CovidcastRow, things that may or may not belong in/with the testing base class, semantics around handling of some values, and some readability stuff.  i also went a little crazy with the underscores in the date ints.

i am also very happy to get on a zoom to pair up to discuss any or all of my comments.

i must say for the record that i dont love the type annotations everywhere, especially because we are not doing any real rigorous checking or enforcement of them, but i will assume your IDE is telling you that they all check out...  i feel a bit like its a plague of locusts coming to descend upon me and maybe i cant get away from it

krivard commented on integrations/client/test_delphi_epidata.py:
```diff
@@ -70,10 +68,11 @@ def test_covidcast(self):
       expected = [
-        self.expected_from_row(row_latest_issue),
-        self.expected_from_row(rows[-1])
+        row_latest_issue.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields),
```
+1; I propose `as_api_compatibility_dict`

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -154,32 +154,25 @@ def test_csv_format(self):
-        row.sample_size
-      ]) + "\n"
+
+    # TODO: This is a mess because of api.php.
```
+1, this PR is a big improvement

or maybe we want something like this?

```suggestion
    # The column ordering is a mess because of api.php.
```

dshemetov replied:
I'm gonna change it to "hardcoded mess" to be more clear 😸

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -200,7 +195,9 @@ def test_fields(self):
-    response, expected = self.request_based_on_row(row, fields='time_value,geo_value')
+    response = self.request_based_on_row(row, fields='time_value,geo_value')
+
+    expected = row.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields)
```
suggestion: another `as_api_compatibility_dict`

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -234,18 +230,15 @@ def test_location_wildcard(self):
-      self.expected_from_row(r) for r in rows[:3]
-    ]
-
+    expected = [row.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields) for row in rows[:3]]
```
suggestion: another `as_api_compatibility_dict`

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -254,18 +247,16 @@ def test_time_values_wildcard(self):
-    expected_time_values = [
-      self.expected_from_row(r) for r in rows[:3]
-    ]
+    expected = [row.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields) for row in rows[:3]]
```
suggestion: another `as_api_compatibility_dict`

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -274,18 +265,16 @@ def test_issues_wildcard(self):
-    expected_issues = [
-      self.expected_from_row(r) for r in rows[:3]
-    ]
+    expected = [row.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields) for row in rows[:3]]
```
suggestion: another `as_api_compatibility_dict`

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -294,12 +283,10 @@ def test_signal_wildcard(self):
-    expected_signals = [
-      self.expected_from_row(r) for r in rows[:3]
-    ]
+    expected_signals = [row.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields) for row in rows[:3]]
```
suggestion: another `as_api_compatibility_dict`

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -314,35 +301,33 @@ def test_geo_value(self):
-    expected_counties = [
-      self.expected_from_row(r) for r in rows[:3]
-    ]
+    expected = [row.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields) for row in rows[:3]]
```
suggestion: another `as_api_compatibility_dict`

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -352,12 +337,10 @@ def test_location_timeline(self):
-    expected_timeseries = [
-      self.expected_from_row(r) for r in rows[:3]
-    ]
+    expected_timeseries = [row.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields) for row in rows[:3]]
```
suggestion: another `as_api_compatibility_dict`

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -383,15 +366,16 @@ def test_unique_key_constraint(self):
-    response, expected = self.request_based_on_row(row)
-    expected.update(stderr=None, sample_size=None)
+    response = self.request_based_on_row(row)
+    expected = row.as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields)
```
suggestion: another `as_api_compatibility_dict`

krivard commented on integrations/server/test_covidcast.py:
```diff
@@ -405,18 +389,19 @@ def test_temporal_partitioning(self):
     # make the request
-    response, expected = self.request_based_on_row(rows[1], time_values="0-99999999")
+    response = self.request_based_on_row(rows[1], time_values="20000101-30010201")
+    expected = [rows[1].as_dict(ignore_fields=CovidcastRow._api_row_compatibility_ignore_fields)]
```
suggestion: another `as_api_compatibility_dict`

krivard commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,241 @@
+    def time_pair(self):
+        return f"{self.time_type}:{self.time_value}"
+
+
```
+1

krivard commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,241 @@
+    _pandas_dtypes: ClassVar = PANDAS_DTYPES
+
+    @staticmethod
+    def make_default_row(**kwargs) -> "CovidcastRow":
```
+1 on putting a subclass in `test_utils`, that might be cleaner overall

krivard reviewed (commented):
Most of these are tagging places where we could drop in `as_api_compatibility_dict` for `as_dict`, but I also have a concern about signing future-us up for more test revision hell the next time we do a major schema change

dshemetov commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,241 @@
+
+    # Classvars.
+    _db_row_ignore_fields: ClassVar = []
+    _api_row_ignore_fields: ClassVar = ["id", "direction_updated_timestamp", "value_updated_timestamp"]
```
I think this needs to be "epimetric_id"

dshemetov commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,241 @@
+        # Convert time values to ints by default.
+        self.time_value = day_to_time_value(self.time_value) if isinstance(self.time_value, date) else self.time_value
+        self.issue = day_to_time_value(self.issue) if isinstance(self.issue, date) else self.issue
+        self.value_updated_timestamp = day_to_time_value(self.value_updated_timestamp) if isinstance(self.value_updated_timestamp, date) else self.value_updated_timestamp
```
sounds good to me!

dshemetov commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,241 @@
+        return df
+
+    @property
+    def api_row_df(self) -> pd.DataFrame:
```
youre totally right about the property thing, removing that

dshemetov commented on src/acquisition/covidcast/test_utils.py:
```diff
@@ -71,13 +57,3 @@ def params_from_row(self, row, **kwargs):
         ret.update(kwargs)
         return ret
-
-    DEFAULT_MINUS=['time_type', 'geo_type', 'source']
-    def expected_from_row(self, row, minus=DEFAULT_MINUS):
-        expected = dict(vars(row))
-        # remove columns commonly excluded from output
-        # nb may need to add source or *_type back in for multiplexed queries
-        for key in ['id', 'direction_updated_timestamp'] + minus:
-            del expected[key]
-        return expected
```
My thinking was that the `as_api_row_df` and `as_api_row_dict` functions should help with that!

dshemetov replied:
Addressed the test cases issue by having one test example and having the rest of the test cases draw from it.

dshemetov commented on src/acquisition/covidcast/test_utils.py:
```diff
@@ -31,36 +33,20 @@ def tearDown(self):
         # close and destroy conenction to the database
         self._db.disconnect(False)
         del self._db
+        self.localTearDown()
```
I'll reorder it, since having the stub is a nice hint for future developers.

dshemetov commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,241 @@
+    id: Optional[int]
+    direction: Optional[int]
+    direction_updated_timestamp: int
+    value_updated_timestamp: int
```
Keeping so that a database row view is possible, but setting defaults.

dshemetov commented on src/server/endpoints/covidcast_utils/model.py:
```diff
@@ -9,6 +9,28 @@
 from ..._params import SourceSignalSet
 
 
+PANDAS_DTYPES = {
```
Alright, moved to covidcast_row.py until JIT or #1037 are addressed.

dshemetov commented on src/acquisition/covidcast/covidcast_row.py:
```diff
@@ -0,0 +1,241 @@
+        self.issue = day_to_time_value(self.issue) if isinstance(self.issue, date) else self.issue
+        self.value_updated_timestamp = day_to_time_value(self.value_updated_timestamp) if isinstance(self.value_updated_timestamp, date) else self.value_updated_timestamp
+
+    def _sanity_check_fields(self, extra_checks: bool = True):
```
Making an issue to refactor csv_importer to use CovidcastRow more here #1078

dshemetov commented on src/server/endpoints/covidcast.py:
```diff
@@ -328,7 +328,7 @@ def parse_row(i, row):
             "signal": row["signal"],
             "time_value": time_value_to_iso(row["time_value"]) if is_day else row["time_value"],
-            "issue": time_value_to_iso(row["issue"]) if is_day else row["issue"],
+            "issue": time_value_to_iso(row["issue"]) if is_day and row["issue"] is not None else row["issue"],
```
I'll revert it and fix whatever test were broken upstream

dshemetov commented on tests/acquisition/covidcast/test_covidcast_row.py:
```diff
@@ -0,0 +1,103 @@
+            "missing_sample_size": Nans.NOT_MISSING,
+            "issue": 20200202,
+            "lag": 0,
+            "direction": None
```
So I've addressed this by writing a single `expected_df` variable at the top of the class and having all the test cases draw from this example.

dshemetov commented:
Thanks for all the review comments! I believe I have addressed all of them and synced with @melange396 on it yesterday. I rebased and force pushed to clean up the review commit log. Should be good to merge.

krivard reviewed (approved):
Glorious triumph 💯
